File: desktop/src/Qlipq.Core/Queue.cs
namespace Qlipq.Core;

/// <summary>Lifecycle of a clip in the editing queue.</summary>
public enum QueueStatus { Pending, Ready, Editing, Exporting, Done, Error }

/// <summary>
/// Per-clip output overrides (quality), merged over the global <see cref="OutputSettings"/>
/// on export. Every field is nullable so only explicitly-set values are persisted and applied
/// (the C# analog of TypeScript's <c>Partial&lt;OutputSettings&gt;</c>).
/// </summary>
public sealed record OutputOverride
{
    public QualityMode? QualityMode { get; set; }
    public QualityPreset? QualityPreset { get; set; }
    public int? Crf { get; set; }
    public int? VideoBitrateKbps { get; set; }
    public string? EncoderPreset { get; set; }
    public VideoCodecChoice? VideoCodec { get; set; }
    public ContainerFormat? Container { get; set; }
    public int? Fps { get; set; }
    public int? MaxHeight { get; set; }
    public int? AudioBitrateKbps { get; set; }
}

/// <summary>A recording tracked in the queue, with any parsed metadata and edit state.</summary>
public sealed record QueueItem
{
    public string Id { get; set; } = "";

    /// <summary>Absolute path on disk.</summary>
    public string Path { get; set; } = "";

    public string FileName { get; set; } = "";

    /// <summary>ISO timestamp of when it entered the queue.</summary>
    public string AddedAt { get; set; } = "";

    public QueueStatus Status { get; set; } = QueueStatus.Pending;

    /// <summary>ISO timestamp parsed from the filename, if any.</summary>
    public string? RecordedAt { get; set; }

    /// <summary>Scene/game label parsed from the filename, if any.</summary>
    public string? Source { get; set; }

    /// <summary>Probed media info, populated lazily when the clip is opened.</summary>
    public MediaInfo? Media { get; set; }

    /// <summary>File size in bytes, read from the filesystem.</summary>
    public long? FileSizeBytes { get; set; }

    /// <summary>ISO timestamp of the file's last modification, from the filesystem.</summary>
    public string? FileModifiedAt { get; set; }

    /// <summary>Clip duration in seconds, probed lazily in the background for the queue.</summary>
    public double? DurationSec { get; set; }

    /// <summary>Working edit spec, persisted so re-opening a clip restores progress.</summary>
    public EditSpec? Edit { get; set; }

    /// <summary>Per-clip output overrides (quality), merged over the global defaults on export.</summary>
    public OutputOverride? OutputOverride { get; set; }

    /// <summary>Free-form labels for filtering the queue.</summary>
    public List<string>? Tags { get; set; }

    /// <summary>Where the last successful export was written.</summary>
    public string? ExportPath { get; set; }

    public string? Error { get; set; }
}

File: apps/desktop/crates/qlipq-core/src/queue.rs
use serde::{Deserialize, Serialize};

use crate::config::{ContainerFormat, QualityMode, QualityPreset, VideoCodecChoice};
use crate::edit_spec::EditSpec;
use crate::media::MediaInfo;

/// Lifecycle of a clip in the editing queue.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum QueueStatus {
    Pending,
    Ready,
    Editing,
    Exporting,
    Done,
    Error,
}

/// Per-clip output overrides, merged over the global [`OutputSettings`](crate::config::OutputSettings)
/// on export. Every field is optional so only set values persist/apply.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputOverride {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub quality_mode: Option<QualityMode>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub quality_preset: Option<QualityPreset>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub crf: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub video_bitrate_kbps: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub encoder_preset: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub video_codec: Option<VideoCodecChoice>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub container: Option<ContainerFormat>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fps: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_height: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub audio_bitrate_kbps: Option<i64>,
}

/// A recording tracked in the queue, with any parsed metadata and edit state.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueueItem {
    pub id: String,
    pub path: String,
    pub file_name: String,
    /// ISO timestamp of when it entered the queue.
    pub added_at: String,
    pub status: QueueStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recorded_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub media: Option<MediaInfo>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file_size_bytes: Option<i64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub file_modified_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub duration_sec: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub edit: Option<EditSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_override: Option<OutputOverride>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub export_path: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}
